### ISSUE TITLE
Visibility question

### ISSUE THREAD

**Issue by alexandru** [JCTools/JCTools#224]
I am trying to build an async queue, that (asynchronously) blocks the producer until the queue has room in it, or that waits on the consumer side until the queue has items in it. 

And depending on the user-provided configuration, I'm using these queues from JCTools:

- `MpmcArrayQueue`
- `MpscArrayQueue`
- `SpmcArrayQueue`
- `SpscArrayQueue`
- `MpmcAtomicArrayQueue`
- `MpscAtomicArrayQueue`
- `SpmcAtomicArrayQueue`
- `SpscAtomicArrayQueue`
- `MpscUnboundedArrayQueue`
- `SpscUnboundedArrayQueue`
- `MpscUnboundedAtomicArrayQueue`
- `SpscUnboundedAtomicArrayQueue`
- `java.util.concurrent.ConcurrentLinkedQueue` (for unbounded, multi-consumer scenarios)

Yes, all of them :-)

The waiting part is however tricky, but I arrived to an algorithm that might be correct, but I'm not well versed in the internals of JCTools, so I need help.

Given this shared state:

```java
queue = new MpscUnboundedArrayQueue() // whatever queue from that list
await = new AtomicReference<Promise<Unit>>(null) // using Scala's Promise
```

The consumer does this (sample code also below):

1. tries to pull from the queue
2. if queue was empty, gets the `Promise` from the `await` atomic, or creates a new `Promise`
3. tries to pull from the queue again, to avoid race condition (problematic)
4. if queue was still empty, then register a listener in that `Promise` and go to sleep (n.b. the `Promise` is essentially a second atomic reference with listeners in it)

The producer then does this:

1. push event in the queue
2. reads the `await` atomic reference
3. if `await` is non-null then set `await` to `null` and completes the promise (thus notifying sleeping consumers)

But I don't know if this algorithm is correct. The danger is a visibility issue, on the consumer side at step 3 above. Due to a lack of experience I don't see clear memory boundaries and I'm unfamiliar with the implementations of those queues. Am I right in thinking that I may have a visibility issue?

Worst case scenario I can timeout that future, but that implies the use of a `SchedulerService` and I need to be sure that it's absolutely needed.

Sample code for the consumer (for simplicity working with unbounded queues):

```java
Future<A> pull() {
  // step 1 (try pull)
  var a = queue.poll();
  if (a != null) {
    return Future.successful(a);
  }

  // step 2 (register promise)
  var ref = await.get();
  while (ref == null) {
    val p = new Promise<Unit>();
    if (await.compareAndSet(null, p))
      ref = p;
    else
      ref = await.get();
  }

  // step 3 (try again) - can we have a problem here ?!?
  a = queue.poll();
  if (a != null) {
    return Future.successful(a);
  }

  // step 4 (sleep, then retry)
  return await.future.flatMap(u -> pull());
}
```

N.B. basically I hope that CAS-loop at step 2 is enough to make whatever was published in the `queue` visible. If not my code is screwed 🙂

Sample code for the producer (for simplicity working with unbounded queues):

```java
void push(A a) {
  // step 1 (push)
  queue.push(a);
  // step 2 (inspect listeners)
  val ref = await.get();
  // step 3 (notify sleeping consumers)
  if (ref != null) ref.complete(unit)
}
```

Also, does it matter if we talk about `offer` / `relaxedOffer` and `poll` / `relaxedPoll`?

**Comment by franz1981** [comment]
> Also, does it matter if we talk about offer / relaxedOffer and poll / relaxedPoll?

The most important thing in the offer logic is that you need to have a full memory barrier right after queue.push(a), before `// step 2 (inspect listeners)`.
The sleeping mechanics is well explained here: https://www.scylladb.com/2018/02/15/memory-barriers-seastar-linux/
TLDR:
In the producer’s case, if the `(inspect listeners)` is moved ahead of the queue push, then the producer may miss a consumer that has just gone to sleep.
That means that any single producer queue (ie that are using just putOrdered operations on offer side) 
need to add an `Unsafe::fullFence` there:
```
void push(A a) {
  // step 1 (push)
  queue.push(a);
  UnsafeAccess.UNSAFE.fullFence();
  //any load after queue.push can't move ahead the publication of a 
  // step 2 (inspect listeners)
  val ref = await.get();
  // step 3 (notify sleeping consumers)
  if (ref != null) ref.complete(unit)
}
```

**Comment by alexandru** [comment]
Damn, I was hoping to get away with just an `await.get()`.

How expensive is that full fence? And can we still do that in Java 11?

In my testing, on my own computer, it seems to work (maybe I'm not testing well), so given the hypothesis that it works in general, another possibility is to just timeout that future after 50 ms or so 🙂

**Comment by franz1981** [comment]
> How expensive is that full fence? And can we still do that in Java 11?
I have built a version that is capable to achieve about 100 M msg/sec vs 400 M msg/sec single producer just to give you the magnitude of having that full barrier.
That's why the ScyllDb guys has gone so far by using other means to have a similar effects without bothering the producer side.
That barrier is not avoidable AFAIK if you don't want to risk to miss some wakeup: having a burst bench make it easier to detect that it's working in my experience.
With burst bench I mean something like: https://github.com/real-logic/benchmarks/blob/master/src/main/java/uk/co/real_logic/benchmarks/latency/ManyToOneConcurrentArrayQueueBenchmark.java
I hope that has helped :+1:

**Comment by alexandru** [comment]
@franz1981 thanks, this is very useful

**Comment by alexandru** [comment]
@franz1981 so in case I'm using the multi-producer versions in JCTools, do I still need the fence, or how could I check it?

I guess I could try that burst bench you mentioned and try to reproduce problems first.

**Comment by franz1981** [comment]
If you use most of the multi-producer versions in JCtools I suggest to use poll on consumer side (instead of relaxedPoll) and it would save you to add a fullFence right after offering to the queue.

**Comment by alexandru** [comment]
Thanks a lot for the info @franz1981, will close this issue now.
